R/teal_slice.R

```diff
@@ -0,0 +1,431 @@
+            args <- make_args()
+          } else {
+            for (iiii in seq_along(subsubitem)) {
+              subsubsubitem <- subsubitem[[iiii]]
+              if (is.bottom(subsubsubitem)) {
+                args$dataname <- names(x)[i]
+                args$varname <- names(subsubitem)[iiii]
+                args$selected <- subsubsubitem$selected
+                args$keep_na <- subsubsubitem$keep_na
+                args$keep_inf <- subsubsubitem$keep_inf
+                args$datalabel <- names(item)[ii]
+                args$target <- names(subitem)[iii]
+                slices[[length(slices)+1]] <- as.teal_slice(Filter(Negate(is.null), args))
+                args <- make_args()
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  do.call(filter_settings, slices)
+}
+
+
+# subset method for teal_slices
+#' @export
+#' @rdname teal_slice
+#' @keywords internal
+#'
+`[.teal_slices` <- function(x, i) {
+  if (missing(i)) i <- seq_along(x)
+  if (length(i) == 0L) return(x[0])
+  if (is.logical(i) & length(i) > length(x)) stop("subscript out of bounds")
+  if (is.numeric(i) & max(i) > length(x)) stop("subscript out of bounds")
+  if (is.character(i) & !all(is.element(i, names(x)))) stop("subscript out of bounds")
+
+  y <- NextMethod("[")
+  attrs <- attributes(x)
+  attrs$names <- attrs$names[i]
+  attributes(y) <- attrs
+  excludes <- unique(unlist(vapply(y, function(ts) ts[["dataname"]], character(1L))))
+  attr(y, "exclude") <- Filter(Negate(is.null), attr(x, "exclude")[excludes])
+  y
+}
+
+
+# concatenate method for teal_slices
+#' @export
+#' @rdname teal_slice
+#' @keywords internal
+#'
+c.teal_slices <- function(...) {
+  x <- list(...)
+  excludes <- lapply(x, attr, "exclude")
+  names(excludes) <- NULL
+  excludes <- unlist(excludes, recursive = FALSE)
+  excludes <- excludes[!duplicated(names(excludes))]
+  count_types <- lapply(x, attr, "count_type")
+  count_types <- unique(unlist(count_types))
+  checkmate::assert_string(count_types)
+
+  do.call(filter_settings, c(unlist(x, recursive = FALSE), list(exclude = excludes, count_type = count_types)))
+}
+
+
+# print method for teal_slices
+#' @export
+#' @rdname teal_slice
+#' @keywords internal
+#'
+print.teal_slices <- function(x, ...) {
+  f <- attr(x, "exclude")
+  ct <- attr(x, "count_type")
+  for (i in seq_along(x)) {
+    ind <- names(x)[i]
+    if (is.null(ind)) ind <- sprintf("[[%d]]", i)
+    cat(ind, "\n")
+    print(x[[i]], ...)
+  }
+  cat("\nnon-filterable variables:")
+  if (is.list(f) & length(f) == 0L) {
+    cat(" none\n")
+  } else {
+    cat("\n")
+    for (i in seq_along(f)) {
+      cat(sprintf(" $ %s: %s", names(f)[i], toString(f[[i]])), "\n")
+    }
+  }
+  cat(sprintf("\ncount type: %s", ct), "\n")
+}
+
+
+# get slices where feature is value
+#' @rdname teal_slice
+#' @keywords internal
+#'
+extract_by_feat <- function(tss, feature, value) {
```

Comment:
What about 'get' instead of 'extract'? I'm also thinking something similar to purrr for the names of these functions.

- `extract_feat` -> `get_feature`
- `extract_by_feat` -> `get_feature_val`
- `extract_fun` -> `get_feature_lgl`
- `extract_fun_s` -> `get_feature_lgl_str`

Reply:
The names of these functions are absolutely up for duscussion.

Reply:
Some history:
`extract_by_feat` was created as a Filter wrapper for the simple purpose of `get slices where this field has this value"
`extract_fun` was later added because information was stored in nested lists, so `extract_by_feat` could not directly refer to a field and things like `extras$target == "y"` had to be evaluated in a `teal_slice`
`extract_fun_s` was then added because programmatic construction of an expression is not trivial, to I went to building a string and converting that to a call within the `extract*` function

Even though there are no nested fields at the moment, the latter still holds true. Therefore, I think subsetting `teal_slices` with a string forming a logical predicate can cover all our needs and the other subsetting functions can be dropped.

How about this:
1. remove `extract_fun` and use the `extract_fun_s` in all cases
2. rename `extract_fun_s` to `slices_which` or `slices_where` (I prefer the former)
3. remove `extract_by_feat` and use `slices_which` in all cases
4. rename `feature` to `field` or `element` `extract_feat` to `slices_field`/`slices_element`